features/375-static-cdn/design.md

```diff
@@ -0,0 +1,140 @@
+# Static Asset Hosting
+
+## Background
+Developer would want to host static content (e.g. single-page-application files) on Skygear and serve them through
+high performance CDN.
+
+## Objectives
+- Developers should be able to upload static content to be hosted at specific path.
+- Hosted static content should be served & cached with CDN if possible.
+- SPA style routing should be supported.
+
+## Design
+
+### Static Deployment
+
+A new type of deployment, `static`, would be supported. This deployment type
+supports configurating a deployment context (`context`) and fallback absolute
+path (`fallback`).
+
+When `static` deployment type is used:
+- The deployment context would be persisted to object storage
+    - symlinks are resolved at upload time
+    - `.skyignore` can be used to exclude files
+- A request with matching path would be routing using static deployment routing
+  algorithm (described below).
+- If a file is matched, the gateway would proxy the request to the backing
+  object storage file.
+- If no file is matched, a generic 404 page would be returned.
+- The response would have appropiate cache control headers, as configured by
+  developer.
+
+### Static Deployment Routing
+
+When a request matched a `static` deployment, only its path would be considered
+during routing. The request would be routed as followed:
+
+1. Let `U` be the request URI, let `P` be the routing path without trailing slashes.
+2. If file `P` exists, return `P`.
+3. If file `Q` = `P/index.html` exists, return `Q`.
+4. If `fallback` is configured, set `U` = `U` with path `fallback` and route again.
+5. Otherwise, no file is matched.
+
+This procedure is similar to nginx directives:
+```
+index index.html
+try_files $uri <fallback>
+```
+
+Sometimes routing procedure is restarted (e.g. in step 4). The routing procedure
+can be performed at most 5 times to prevent infinite loop. In this case, `500
+Internal Server Error` response would be produced.
+
+**Example**:
+
+Suppose developer configured a `static` deployment as follow:
+```yaml
+  - name: assets
+    type: static
+    path: /
+    context: ./static
+    fallback: /index.html
```

Comment:
I think the term `fallback` is a bit misleading. Instead, maybe better call it "index", and also add configuration for 404 (4xx class of error codes) which is commonly needed too?

Reply:
I think `fallback` is actually quite clear on what it does: if the file is not found, fallback to this file. This config is mainly used for 404 handling and SPA routing, and we do not support other 4xx HTTP code handling (it seems to me there's no need to support so for static assets).

Reply:
I think we use follow tradition. It actually took me like 0.5 second to realize what fallback could meant. By tradition I think index is the word used by various services from apache/nginx to similar configuration at s3. Who use fallback in similar services?

Moreover, fallback here serve two purpose: both display index page and 4xx error. Those are definitely different. Most static website / asset hosting will need a different design of index page and 404 not found. In fact it would be even better if it is easy to define a different page for 4xx vs 5xx error.

Reply:
p.s. use 4xx instead of specific 404 here just coz S3 do it.

Reply:
> By tradition, I think index is the word used by various services from apache/nginx to the similar configuration at s3.

This configuration is called error document/page in S3, nginx, and apache. I can change its name to `errorPage` if you find it easier to understand.

> Moreover, fallback serves two purposes: both display index page and 4xx error.

That's not correct. In fact, index page are always served from `index.html` if exists. `fallback` configuration is only for 404.

> In fact it would be even better if it is easy to define a different page for 4xx vs 5xx error.

Static routing would only produce 'not found' errors. Other 4xx errors (e.g. 401/403), is not produced. 5xx errors are not always recoverable, and I don't think there are providers supporting custom 5xx pages.

Reply:
I guess S3 allows us to specify the status code because S3 objects may not be publicly accessible. So the status code could be 403 instead of 404.

Reply:
@kiootic, in that case, I think we should have `error` and `index`; We could have `index.html` as default but it should be changeable.

Reply:
@kiootic in that case I think we should have, ideally:

* An option to change the index page's name. We should default it to `index.html`, but it should be changeable in case some poor dev is stuck with something else...
* We should have an error page for 404
* Ideally, we should also have an error page config for 5xx errors. 5xx errors are not recoverable, but for UX purposes, say Skygear screw up something and need to display 5xx, some business may prefer to have its own "We're sorry" page, instead of a Skygear Branded or a plain HTML text of 5xx error page. However, if there is some reason make it difficult to implement I can live without it for now.

Reply:
> An option to change the index page's name. We should default it to index.html, but it should be changeable in case some poor dev is stuck with something else...

I don't think it is a required feature. Using `index.html` is a convention common enough that there should be no need cater for this at the moment. We can revisit this when there's concrete scenario that request this feature.

> We should have an error page setting for 404

I'll change the `fallback` config to `errorPage` instead then.

> Ideally, we should also have an error page config for 5xx errors. 5xx errors are not recoverable, but for UX purposes, say Skygear screw up something and need to display 5xx, some business may prefer to have its own "We're sorry" page, instead of a Skygear Branded or a plain HTML text of 5xx error page. However, if there is some reason make it difficult to implement I can live without it for now.

If a 5xx error is produced when serving static asset, it is very likely that serving another static asset (i.e. custom error page) would fail too. Therefore it is very likely that the custom 5xx error page would not be served for all 5xx errors, and developers would be confused on why sometimes the custom 5xx error page is not used.

Reply:
@kiootic let's stick with a configuration for `index` and one for `error` then. I think there are not much overhead to maintain an `index` configuration option, this configuration is also very common (S3, apache, nginx, anything I can think of have this option); Given there are also known usage of this (legacy one like `index.htm`, `index.xhtml`, or sub-folder, etc), I don't see why we shouldn't do it now vs have to re-visit it in future.

Good point on 5xx error might not served at all so render the feature not very useful.

Reply:
There are some additional considerations if we want to include this index page configuration:
- Do we want to support multiple custom index page file name? (e.g. use `index.html` if exist, otherwise `index.htm`)
- Do we want to include a default set of common index page names?

Reply:
I think default to `index.html` is good enough and let the user pick those obsoleted one is good enough. Set is a bit too much in our case?

However if we're going deep into it, one common use case is support multiple languages index page depends on Accept-Language and IP address geo... however I think that's a rabbit hole we don't want to dig into it yet...?